### PR TITLE
feat(util): Support multiple AWS key sources

### LINKED ISSUE #13: AWS::Init should use multiple fallbacks to find credentials
Right now AWS::Init uses only AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY to find credentials.

However, there are other ways to do it.
The order is:
1. To check environment variables  https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
2. If it's not enough - check for credential file https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html
   (need to use some of the env vars to know what to look for)
3. if it's not enough - check the instance metadata (prod instance)

(one can check how aws cli checks for credentials via `aws --debug s3api list-buckets` for example) 

To do (2) - we can use `io::ini::Parse` method from `line_reader.h`.
To do 3 -  one can use  https://github.com/aws/amazon-ec2-metadata-mock and then of course to do the final testing on ec2 instance.

### PR DESCRIPTION
**Pull request by dranikpg**
Fixes #13

### CODE REVIEW

File: util/cloud/aws.cc [hunk 1]
```diff
@@ -114,6 +123,119 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+  http::Client http_client{pb};
+  ec = http_client.Connect("127.0.0.1", "1338");
+  if (ec)
+    return std::nullopt;
+
+  // TODO: fix role name
+  const char* PATH = "/latest/meta-data/iam/security-credentials/baskinc-role";
+  h2::request<h2::empty_body> req{h2::verb::get, PATH, 11};
+  h2::response<h2::string_body> resp;
+  req.set(h2::field::host, http_client.host());
```
[the remark this answers is not in the document]
Comment (romange) [reply util/cloud/aws.cc:199]: Send me an email with your public ssh key - i will provide you with access to  an ec2 instance. Will be easier to discuss.

File: util/cloud/aws.cc [hunk 2]
```diff
@@ -114,6 +123,119 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+      keys.session_token = session_token;
+    }
+    return keys;
+  }
+  return std::nullopt;
+}
+
+// Try reading AwsKeys from credentials file.
+std::optional<AwsKeys> GetKeysFromFile() {
+  const char* home_folder = getenv("HOME");
```
Comment (romange) [inline util/cloud/aws.cc:149]: if AWS_SHARED_CREDENTIALS_FILE is defined you do not need to check HOME

File: util/cloud/aws.cc [hunk 3]
```diff
@@ -114,6 +123,119 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+std::optional<AwsKeys> GetKeysFromMetadata() {
+  error_code ec;
+  ProactorBase* pb = ProactorBase::me();
+  CHECK(pb);
+
+  // TODO: Replace mock path with real. Make mockable path?
+  http::Client http_client{pb};
+  ec = http_client.Connect("127.0.0.1", "1338");
+  if (ec)
+    return std::nullopt;
```
[the remark this answers is not in the document]
Comment (romange) [reply util/cloud/aws.cc:193]: yes

File: util/cloud/aws.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (romange) [reply util/cloud/aws.cc:199]: if you curl the parent path like this 
`curl  http://169.254.169.254/latest/meta-data/iam/security-credentials/` you get the name of the profile. so you just need 2 hops.

File: util/cloud/aws.cc [hunk 4]
```diff
@@ -114,6 +123,119 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+    keys.access_key = it->second["aws_access_key_id"];
+    keys.secret_key = it->second["aws_secret_access_key"];
+    keys.session_token = it->second["aws_session_token"];
+    return keys;
+  }
+  return std::nullopt;
+}
+
+// Try getting AwsKeys from instance metadata.
+std::optional<AwsKeys> GetKeysFromMetadata() {
```
[the remark this answers is not in the document]
Comment (romange) [reply util/cloud/aws.cc:184]: yes, we should probably at least take it from metadata. That means you would need to restructure  `AwsKeys`.

**Comment by dranikpg** [comment]
I still don't really like the way error handling works 🙂 A few logs would probably make finding fail causes easier

**Comment by romange** [comment]
1. Where you would add an error log and for what reason?
2. You may propagate error_code via io::Result but it's important to
understand what you consider an error case

**Comment by dranikpg** [comment]
1. To see why `Aws::Init()` failed (wrong file, wrong json/ini format, no iam role attached) and what it tried. This would make troubleshooting a lot easier. Currently if functions just like a yes/no black box.
2. Programmatic errors are probably not that important, because the end consumer is an assert.

**Comment by romange** [comment]
In that case just LOG(ERROR) where you are already committed to some path.
As you said, you found the file but failed to parse, or found the metadata
but it's missing role.

File: util/cloud/aws.cc [hunk 5]
```diff
@@ -114,6 +122,196 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+  if (!contents) {
+    LOG(ERROR) << "Failed to parse ini file:" << full_path;
+    return std::nullopt;
+  }
+
+  return contents.value();
+}
+
+// Try filling AwsConnectionData with data from config file.
+void GetConfigFromFile(AwsConnectionData& cd, const char* profile) {
```
Comment (romange) [inline util/cloud/aws.cc:181]: Google style-guide: output arguments should be last one and passed by a pointer. I.e here 'FillConfigFromFile(const char*, Aws...* cd)'

File: util/cloud/aws.cc [hunk 6]
```diff
@@ -114,6 +122,196 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+  }
+
+  if (profile != "default") {
+    LOG(ERROR) << "Failed to find profile:" << profile << " in credentials";
+  }
+  return std::nullopt;
+}
+
+// Make simple GET request on path and return body.
+std::optional<std::string> MakeGetRequest(http::Client& http_client, boost::string_view path) {
```
Comment (romange) [inline util/cloud/aws.cc:230]: Here the same.

File: util/cloud/aws.cc [hunk 7]
```diff
@@ -114,6 +122,196 @@ inline std::string_view std_sv(const ::boost::beast::string_view s) {
...
+  req.set(h2::field::host, http_client.host());
+
+  std::error_code ec = http_client.Send(req, &resp);
+  if (ec || resp.result() != h2::status::ok)
+    return std::nullopt;
+
+  return resp.body();
+}
+
+void GetConfigFromMetadata(AwsConnectionData& cd, http::Client& http_client) {
```
Comment (romange) [inline util/cloud/aws.cc:242]: Same here, i would also swap the order because 'cd' is the real output argument

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]

**Pull request merged**